modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1245 @@
+
+      :arg ch: A channel to write to.
+    */
+    proc writeThis(ch: channel) throws {
+      _enter();
+      
+      ch <~> "[";
+
+      var first = true;
+
+      for x in this {
+        if !first then ch <~> ", ";
+        else first = false;
+        ch <~> x;
+      }
+
+      ch <~> "]";
+
+      _leave();
+    }
+
+    /*
+      Returns `true` if this unrolledLinkedList contains zero elements.
+
+      :return: `true` if this unrolledLinkedList is empty.
+      :rtype: `bool`
+    */
+    proc const isEmpty(): bool {
+      return size == 0;
+    }
+
+    /*
+      The current number of elements contained in this unrolledLinkedList.
+    */
+    inline proc const size {
+      var result = 0;
+
+      on this {
+        _enter();
+        result = _size;
+        _leave();
+      }
+
+      return result;
+    }
+
+    /*
+      Returns the unrolledLinkedList's legal indices as the range ``0..<this.size``.
+
+      :return: ``0..<this.size``
+      :rtype: `range`
+    */
+    proc indices {
+      return 0..<this.size;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy-initialized var of 
+      each of the elements contained in this unrolledLinkedList.
+
+      :return: A new DefaultRectangular array.
+    */
+    proc const toArray(): [] eltType {
+      if isNonNilableClass(eltType) && isOwnedClass(eltType) then
+        compilerError("toArray() method is not available on a 'unrolledLinkedList'",
+                      " with elements of a non-nilable owned type, here: ",
+                      eltType:string);
+
+
+      var result: [0..#_size] eltType;
+
+      on this {
+        _enter();
+
+        var i = 0;
+        for x in this {
+          result[i] = x;
+          i += 1;
+        }
+
+        _leave();
+      }
+
+      return result;
+    }
+
+  } // End record "unrolledLinkedList".
+
+  /*
+    Clear the contents of this unrolledLinkedList, then extend this now empty unrolledLinkedList with the
+    elements contained in another unrolledLinkedList.
+
+    .. warning::
+
+      This will invalidate any references to elements previously contained in
+      `lhs`.
+
+    :arg lhs: The unrolledLinkedList to assign to.
+    :arg rhs: The unrolledLinkedList to assign from. 
+  */
+  proc =(ref lhs: unrolledLinkedList(?t, ?), rhs: unrolledLinkedList(t, ?)) {
+    lhs.clear();
+    lhs.extend(rhs);
+  }
+
+  /*
+    Returns `true` if the contents of two unrolledLinkedLists are the same.
+
+    :arg a: A unrolledLinkedList to compare.
+    :arg b: A unrolledLinkedList to compare.
+
+    :return: `true` if the contents of two unrolledLinkedLists are equal.
+    :rtype: `bool`
+  */
+  proc ==(a: unrolledLinkedList(?t, ?), b: unrolledLinkedList(t, ?)): bool {
+    if a.size != b.size then
+      return false;
+
+    // TODO: Optimization here. This is O(N^3)
+    for i in 0..#(a.size) do
+      if a[i] != b[i] then
+        return false;
```

Comment:
This would be a good thing to check the performance of.  I think you could write

```chapel
for (ael, bel) in zip(a, b) do
  if ael != bel then
    return false;
```

and it should step through the two O(n) iterators.

Reply:
I think the existing algorithm is only O(n^2), since the loop is done n times, and each of `a[i]` and `b[i]` are O(n), but they're not multiplicative.